en/docs/extended_topics/user_op.md

```diff
@@ -1,414 +1,816 @@
-           .Output("out") \
-           .Build().RemoteBlobList()[0]
-# flow.user_op_builder(op_name)  -- 生成一个 op_name 作为其op的名称的op
-#     .Op("Relu")             -- 这个op的类型是Relu
-#     .Input("in", [input_blob]) -- 这个op有一个input_arg_name 为 "in"，并且对应的输入blob列表是[input_blob]，表示in只对应一个blob
-#     .Output("out")             -- 这个op有一个output_arg_name 为 "out"，默认out对应的blob数量是1
-#     .Build()                   -- 生成该op的python wrapper
-#     .RemoteBlobList()[0]       -- 返回该op的唯一输出blob
+    op = (
+        flow.user_op_builder("op_myrelu")
+        .Op("myrelu")
+        .Input("in", [input_blob])
+        .Output("out")
+        .Build()
+    )
 ```
 
-## 高级特性
-到现在为止，我们已经完成`Relu` op的构建。当然，Relu Op是一个比较简单的Op，如果我们需要构建一个比较复杂的Op，就需要使用一些额外的高级特性来协助我们。
+This object contains `Op`, `Input` and etc which are used to encapsulate custom op. Details explanation are as follows:
 
-### Op Registration
+* `Op("myrelu") `: The parameter must be the `op_type_name` from the previous C++ registration which OneFlow uses to find the registered op type and instantiate the op.
 
-#### Attribute
+* `Input("in", [input_blob]) `: Corresponds to `Input` when op is registered in C++ that the first parameter must be the same as the string set by `Input` when op is registered in C++. The second parameter is the blob of the input which is a `list`. Because an op allows multiple inputs.
 
-有的Op需要有配置属性，例如`Conv` op需要配置其`padding`的方式、`Reshape` op需要配置一个`tensor shape`。当Op被添加到Graph中时，我们就需要给这些属性设置合理的值了。
+* `Output("out") `: Corresponds to `Output` when op registered in C++.
 
-在OneFlow中，你可以在注册Op时指明其需要的属性Attr，此处我们以`Reshape`  op为例：
+* `Build`：After the above settings are complete, call `Build` to get the Python wrapper from the custom op.
 
-```cpp
-REGISTER_USER_OP("Reshape")
-    .Input("in")
-    .Output("out")
-    .Attr("shape", UserOpAttrType::kAtShape)
-    .SetShapeInferFn([](user_op::InferContext* ctx) -> Maybe<void> {
-      const Shape* in_shape = ctx->Shape4ArgNameAndIndex("in", 0);
-      Shape* out_shape = ctx->Shape4ArgNameAndIndex("out", 0);
-      Shape conf_shape = ctx->GetAttr<Shape>("shape");
-      CHECK_EQ(in_shape->NumAxes(), conf_shape.NumAxes());
-      *out_shape = conf_shape;
-      return Maybe<void>::Ok();
-    });
+The following code will get the blob of the custom op:
+
+```python
+return op.InferAndTryRun().SoleOutputBlob()
 ```
 
-上述代码中，我们为`Reshape` op配置了一个attribute，名字为`shape`，其类型是`UserOpAttrType::kAtShape`。
+`InferAndTryRun` completes the derivation and returns `UserOp`. If the returned blob has only one output. We cab use `SoleOutputBlob` to get the unique output. Otherwise use `RemoteBlobList` to get a list of multiple blobs.
 
-在OneFlow中，我们目前支持了如下几种AttrType：
+So far, we have built the `myrelu` which is a relatively simple op. But if we need to build a more complex op, we should use some additional features in the registration process.
+We'll introduce from op registration, kernel registration, gradient registration and Python layer wrapping.
```

Comment:
We'll introduce it  from the aspects of op registration, kernel registration, gradient registration and Python layer wrapping.